internal/service/wallet.go

```diff
@@ -226,37 +223,122 @@ func (s *walletService) GetWalletTransactions(ctx context.Context, walletID stri
-		TransactionReason: types.TransactionReasonFreeCredit,
-		ReferenceType:     types.WalletTxReferenceTypeRequest,
-		ReferenceID:       types.GenerateUUIDWithPrefix(types.UUID_PREFIX_WALLET_TRANSACTION),
+		TransactionReason: req.TransactionReason,
+		ReferenceType:     referenceType,
+		ReferenceID:       referenceID,
 		ExpiryDate:        req.ExpiryDate,
+		IdempotencyKey:    idempotencyKey,
 	}
 
-	if req.PurchasedCredits {
-		if req.GenerateInvoice {
-			creditReq.TransactionReason = types.TransactionReasonPurchasedCreditInvoiced
-		} else {
-			creditReq.TransactionReason = types.TransactionReasonPurchasedCreditDirect
-		}
+	// Process wallet credit
+	if err := s.CreditWallet(ctx, creditReq); err != nil {
+		return nil, err
 	}
 
-	if creditReq.ReferenceID == "" || creditReq.ReferenceType == "" {
-		creditReq.ReferenceID = types.GenerateUUIDWithPrefix(types.UUID_PREFIX_WALLET_TRANSACTION)
-		creditReq.ReferenceType = types.WalletTxReferenceTypeRequest
+	return s.GetWalletByID(ctx, walletID)
+}
+
+func (s *walletService) handlePurchasedCreditInvoicedTransaction(ctx context.Context, walletID string, idempotencyKey *string, req *dto.TopUpWalletRequest) (string, error) {
+	// Initialize required services
+	invoiceService := NewInvoiceService(s.ServiceParams)
+	customerService := NewCustomerService(s.ServiceParams)
+	paymentService := NewPaymentService(s.ServiceParams)
+
+	// Retrieve wallet and customer details
+	w, err := s.WalletRepo.GetWalletByID(ctx, walletID)
+	if err != nil {
+		return "", err
 	}
 
-	if err := s.CreditWallet(ctx, creditReq); err != nil {
-		return nil, err
+	customer, err := customerService.GetCustomer(ctx, w.CustomerID)
+	if err != nil {
+		return "", err
 	}
 
-	return s.GetWalletByID(ctx, walletID)
+	var paymentID string
+	err = s.DB.WithTx(ctx, func(ctx context.Context) error {
+		// Create invoice for credit purchase
+		invoice, err := invoiceService.CreateInvoice(ctx, dto.CreateInvoiceRequest{
+			CustomerID:     customer.ID,
+			AmountDue:      req.CreditsToAdd,
+			Currency:       w.Currency,
+			InvoiceType:    types.InvoiceTypeCredit,
+			DueDate:        lo.ToPtr(time.Now().UTC()),
+			IdempotencyKey: idempotencyKey,
+			InvoiceStatus:  lo.ToPtr(types.InvoiceStatusFinalized),
+			LineItems: []dto.CreateInvoiceLineItemRequest{
+				{
+					Amount:      req.CreditsToAdd,
+					Quantity:    decimal.NewFromInt(1),
+					DisplayName: lo.ToPtr("Purchased Credits"),
+				},
+			},
+			PaymentStatus: lo.ToPtr(types.PaymentStatusPending),
+		})
+		if err != nil {
+			return err
+		}
+
+		// Create payment for the invoice
+		// Process : true will process the payment and update the invoice payment status
+		payment, err := paymentService.CreatePayment(ctx, &dto.CreatePaymentRequest{
+			IdempotencyKey:    lo.FromPtr(idempotencyKey),
+			DestinationType:   types.PaymentDestinationTypeInvoice,
+			DestinationID:     invoice.ID,
+			PaymentMethodType: types.PaymentMethodTypeOffline,
+			Amount:            req.CreditsToAdd,
```

Comment:
Need to convert the Credits Amount back to Currency amount
amount in the currency =  number of credits * conversion_rate